mitiq/zne/scaling/identity_insertion.py

```diff
@@ -0,0 +1,134 @@
+
+    Args:
+        input_circuit_depth : Number of moments in the input_circuit
+        scale_factor : Intended noise scaling factor
+
+    Returns:
+        (num_uniform_layers, num_partial_layers) : A tuple of the number of
+        uniform identity layers to be inserted after each moment in the
+        input_circuit and a number of partial layers to be inserted after
+        some random moments to be able to achieve the intended scale factor.
+    """
+    if scale_factor < 1:
+        raise ValueError(
+            f"Requires scale_factor >= 1 but scale_factor = {scale_factor}."
+        )
+
+    # find number of uniform layers
+    num_uniform_layers = int(scale_factor - 1)
+    int_scale_factor = num_uniform_layers + 1
+    if np.isclose(int_scale_factor, scale_factor):
+        return (num_uniform_layers, 0)
+    else:
+        # find partial layers by approximating closest to the desired
+        # scale_factor
+        num_partial_layers = int(
+            input_circuit_depth * (scale_factor - 1 - num_uniform_layers)
+        )
+        return (num_uniform_layers, num_partial_layers)
+
+
+# identity insertion scaling function
+def insert_id_layers(input_circuit: Circuit, scale_factor: float) -> Circuit:
+    """Returns a scaled version of the input circuit by inserting layers of
+    identities.
+
+    Args:
+        input_circuit : Cirq Circuit to be scaled
+        scale_factor : Noise scaling factor as a float
+
+    Returns:
+        scaled_circuit : Scaled quantum circuit via identity layer insertions
+    """
+    # Calculate input circuit depth after checking if it is scalable, remove
+    # terminal measurements
+    _check_scalable(input_circuit)
+    measurements = _pop_measurements(input_circuit)
+    input_circuit_depth = len(input_circuit)
+
+    # find number of uniform and partial layers
+    num_uniform_layers, num_partial_layers = _calculate_id_layers(
+        input_circuit_depth, scale_factor
+    )
+
+    # find list of random moments for partial layers
+    random_moment_indices = np.random.randint(
+        input_circuit_depth, size=num_partial_layers
+    )
+    # figure out if the random list has any repeated indices
+    index_counter = Counter(random_moment_indices)
+
+    # create a layer of identity acting on every qubit in the circuit
+    circuit_qubits = input_circuit.all_qubits()
+    id_layer = Moment(ops.I.on_each(*circuit_qubits))
+
+    # create the scaled circuit
+    scaled_circuit = Circuit()
+    for i in range(input_circuit_depth):
+        scaled_circuit.append(input_circuit[i])
+        # add partial layers if i is in random moment index list
+        if i in random_moment_indices:
+            num_partial_layers_random_moment = index_counter[i]
+            scaled_circuit.append(
+                [id_layer] * num_partial_layers_random_moment
+            )
+        # now insert uniform layers
+        scaled_circuit.append([id_layer] * num_uniform_layers)
```

Comment:
```suggestion
    for i, op in enumerate(input_circuit):
        scaled_circuit.append(op)

        scaled_circuit.append(
            [id_layer] * (num_uniform_layers + index_counter.get(i, 0))
        )
```

What do you think of this?

Reply:
A lot more cleaner ! 

Again, this is due to me sticking to my comfort zone. My first programming class was a for loop over a list using range and that's what I remember when working out a for loop. Perhaps this is a good sign that I should venture out from my comfort zone.

Reply:
Sweet! Code review is a great place to learn new things! Glad I could help 😄.